cachito/web/api_v1.py

```diff
@@ -33,14 +36,54 @@ def get_request(request_id):
     :return: a Flask JSON response
     :rtype: flask.Response
     :raise NotFound: if the request is not found
-    :raise ValidationError: if the request ID is of invalid format
     """
-    try:
-        request_id = int(request_id)
-    except ValueError:
-        raise ValidationError(f'{request_id} is not a valid request ID')
+    return flask.jsonify(Request.query.get_or_404(request_id).to_json())
+
+
+@api_v1.route('/requests/<int:request_id>/download', methods=['GET'])
+def download_archive(request_id):
+    """
+    Download archive of source code.
+
+    :param int request_id: the value of the request ID
+    :return: a Flask send_file response
+    :rtype: flask.Response
+    :raise NotFound: if the request is not found
+    """
     request = Request.query.get_or_404(request_id)
-    return flask.jsonify(request.to_json())
+    # TODO: Verify request has already been processed.
+
+    cachito_shared_dir = flask.current_app.config.get('CACHITO_SHARED_DIR')
+    wait_timeout = flask.current_app.config.get('CACHITO_WAIT_TIMEOUT')
+
+    with tempfile.TemporaryDirectory(prefix='cachito-', dir=cachito_shared_dir) as temp_dir:
+        # Although the cachito_shared_dir volume is required to be the same between celery
+        # workers and the API, they may be mounted at different locations. Use relative
+        # paths to agree on data location within the shared volume.
+        relative_temp_dir = os.path.basename(temp_dir)
+        relative_deps_path = os.path.join(relative_temp_dir, 'deps')
+        relative_app_archive_path = os.path.join(relative_temp_dir, 'app.tar.gz')
+        absolute_deps_path = os.path.join(cachito_shared_dir, relative_deps_path)
+        absolute_app_archive_path = os.path.join(cachito_shared_dir, relative_app_archive_path)
+
+        # Chain tasks
+        chain_result = chain(
+            tasks.fetch_app_source.s(request.repo, request.ref,
+                                     copy_cache=relative_app_archive_path),
+            tasks.fetch_gomod_source.s(copy_cache=relative_deps_path)
+        ).delay()
+        chain_result.wait(timeout=wait_timeout)
+
+        # Generate a tarball containing the application and dependencies source code
+        response_archive_path = os.path.join(temp_dir, 'temp.tar.gz')
+        with tarfile.open(response_archive_path, mode='w:gz') as response_archive:
```

Comment:
It's not a requirement, but I'd prefer if instead of chaining two tasks together and then generating the final tar file on the API, you just had one synchronous task on the worker that provides you with the final tar file to stream to the user. It just feels a bit off to me that part of the work happens on the worker and part of it happens on the API.

Reply:
I did consider creating a new task in the worker to handle this. The problem is that this would be a task that calls other tasks synchronously. This is not recommended as it may cause deadlocks.

Reply:
@lcarva the task doesn't have to call other tasks. The current tasks are merely wrappers around a couple functions. The new task would just call those two functions.